File: Code_Python/filter_egarch_asymmetric.py
# filter_egarch_asymmetric.py
import numpy as np
from scipy.special import gamma

class FilterEGARCHAsymmetric:
    def __call__(self, mu, initial_lambda, phi, kappa, kappa_tilde, nu, returns):
        T = len(returns)

        u = np.zeros(T)
        v = np.zeros(T)
        epsilon = np.zeros(T)
        lamdba = np.zeros(T)
        sigma = np.zeros(T)

        lamdba[0] = initial_lambda
        sigma[0] = np.exp(lamdba[0])

        for t in range(1, T):
            epsilon[t-1] = (returns[t-1] - mu) * np.exp(-lamdba[t-1])
            u[t-1] = np.sqrt((nu + 3) / (2 * nu)) * (((nu + 1) / (nu - 2 + epsilon[t-1] ** 2)) * epsilon[t-1] ** 2 - 1)
            v[t-1] = np.sqrt((nu - 2) * (nu + 3) / (nu * (nu + 1))) * ((nu + 1) / (nu - 2 + epsilon[t-1] ** 2)) * epsilon[t-1]

            lamdba[t] = initial_lambda * (1 - phi) + phi * lamdba[t-1] + kappa * u[t-1] + kappa_tilde * v[t-1]
            sigma[t] = np.exp(lamdba[t])

        epsilon[-1] = (returns[-1] - mu) * np.exp(-lamdba[-1])
        u[-1] = np.sqrt((nu + 3) / (2 * nu)) * (((nu + 1) / (nu - 2 + epsilon[-1] ** 2)) * epsilon[-1] ** 2 - 1)
        v[-1] = np.sqrt((nu - 2) * (nu + 3) / (nu * (nu + 1))) * ((nu + 1) / (nu - 2 + epsilon[-1] ** 2)) * epsilon[-1]

        return sigma, lamdba, epsilon, u, v

File: Code_MATLAB/Filter_EGARCH_asymmetric.m
function [lambda, sigma, epsilon, u, v] = Filter_EGARCH_asymmetric(mu, initial_lambda, phi, kappa, kappa_tilde, nu, returns)
    % Extract the sample size
    T = size(returns, 1);

    % Prefill the variables we are going to track
    u = zeros(T,1);
    v = zeros(T,1);
    epsilon = zeros(T,1);
    lambda = zeros(T, 1);
    sigma = zeros(T, 1);
    
    % Initialize at the unconditional mean of lambda
    lambda_start = initial_lambda;
    sigma_start = exp(lambda_start);

    % Run the Beta-t-EGARCH filter
    for t = 1:T
        if t==1
            sigma(t) = sigma_start;
            lambda(t) = initial_lambda;
            epsilon(t) = (returns(t) - mu)*exp(-lambda(t));
            u(t) = sqrt((nu + 3) / (2 * nu)) * (((nu + 1) / (nu - 2 + epsilon(t)^2)) *epsilon(t)^2 - 1);
            v(t) = sqrt((nu - 2) * (nu + 3) / (nu * (nu + 1))) * ((nu + 1) / (nu - 2 + epsilon(t)^2)) * epsilon(t);
        % Compute residual
        else 
            lambda(t) = initial_lambda * (1 - phi) + phi * lambda(t-1) + kappa * u(t-1) + kappa_tilde * v(t-1);
            sigma(t) = exp(lambda(t));
            epsilon(t) = (returns(t) - mu)*exp(-lambda(t));
            u(t) = sqrt((nu + 3) / (2 * nu)) * (((nu + 1) / (nu - 2 + epsilon(t)^2)) *epsilon(t)^2 - 1);
            v(t) = sqrt((nu - 2) * (nu + 3) / (nu * (nu + 1))) * ((nu + 1) / (nu - 2 + epsilon(t)^2)) * epsilon(t);
        end
        
    end
end
